[DOCS] Revamp Airflow Deployment Pattern (#3963)
great-expectations/great_expectations: docs/deployment_patterns/how_to_run_a_checkpoint_in_airflow.md

</Prerequisites>

> Airflow is a data orchestration tool focused on the easy creation and maintenance of data pipelines through DAGs (directed acyclic graphs) written in Python. The base unit of work done in a DAG is the Operator, and this document shows how to use the `GreatExpectationsOperator` to perform data quality checks in the context of an ELT pipeline.

Reviewer: ```suggestion
Airflow is a data orchestration tool for creating and maintaining data pipelines through DAGs (directed acyclic graphs) written in Python. DAGs complete work through operators, which are templates that each encapsulate a specific type of work. This document explains how to use the `GreatExpectationsOperator` to perform data quality work in an Airflow DAG. 
```

Reviewer: Decided to uncapitalize "operator" when it's not in reference a specific kind of operator (changed my mind on this since the last time I edited you, sorry).

> While Airflow alone is a powerful tool, [Astronomer](https://www.astronomer.io/) provides a CLI and cloud platform to bring your DAGs to the next level by providing simple and efficient ways to create Airflow projects, author and publish DAGs, and easily maintain the Kubernetes engine powering DAG runs. To get started quickly with Astronomer’s CLI for use in the steps below, check out the [Astronomer CLI quick start guide](https://docs.astronomer.io/enterprise/cli-quickstart/).

Reviewer: Idk, seems a bit cheeky to have this in the Great Expectations version of the doc, especially since none of these tools are required to follow the guide. I would expect this information to be in a guide that's similar to what's in their [Cloud Composer setup](https://docs.greatexpectations.io/docs/deployment_patterns/how_to_use_great_expectations_with_google_cloud_platform_and_bigquery#3-install-great-expectations-in-cloud-composer). Do we have plans to write a doc like that for Astronomer?

Author: We are looking to have more reference architecture docs on our end. I can get rid of all the Astronomer language, as I have to update [our GE guide](https://www.astronomer.io/guides/airflow-great-expectations) anyway. So all the Astronomer-specific language can go there.

Reviewer 2: Agreeing with @jwitz, I think it could make sense to have this as a general Airflow guide, and then have an additional guide for Astronomer specifically (leveraging lots of links to this guide so we don't have too much duplicated content).

Reviewer 2: If this is an Airflow-specified guide, I could see including a link to the Astronomer guide here as well. A note like:
"""
This guide focuses on using Great Expectations with Airflow in a self-hosted environment. See here for the guide on using Great Expectations with Airflow from within Astronomer.
"""

Author: Ah, thought I removed this paragraph. Replacing now.

> ## Background

Reviewer: Instead of framing it like a two step process, how about we change this header to "Install the GreatExpectationsOperator" and keep the usage as a separate idea from the setup.

Importing providers into Airflow is an easy two-step process. The first step is either `pip install`-ing the Great Expectations provider in your Airflow environment or adding the provider `airflow-provider-great-expectations` to your `requirements.txt` file in your Airflow project. This file is generated for you if you are using the Astronomer CLI and run `astro dev init`, and the provider is installed on to your Docker image when spinning up Airflow.

> It’s recommended to specify a version in the `requirements.txt` file. To make use of the latest Great Expectations V3 API, you’ll need to specify a version `>=0.1.0`, and if you’re using Astronomer, you’ll need version `>=0.1.1`. In either case, the Great Expectations V3 API provider requires Airflow 2, so if you’re still running Airflow 1.X, upgrade that first.

Reviewer: ```suggestion

To import the GreatExpectationsOperator in your Airflow project, run the following command to install the Great Expectations provider in your Airflow environment: 

```
pip install airflow-provider-great-expectations==<version-number>
```

It’s recommended to specify a version when installing the package. To make use of the latest Great Expectations V3 API, you need to specify a version >= `0.1.0`. 

:::note 

The Great Expectations V3 API requires Airflow 2.0+. If you're still running Airflow 1.x, you need to upgrade to at least 2.0 before using v0.1.0+ of the GreatExpectationsOperator.

:::

```

Reviewer: Ugh GitHub butchered this edit, you'll have to click "edit" on the post to see what the my intent was. 

Removing Astronomer related stuff until I can figure out exactly where I wanna put it.

Author: I can add this edit in on my end.

## Using the Operator

> The second step to use the operator is to import it into the DAG. The DAG file is a python file in the `dags/` folder of your Airflow project. To import the Great Expectations provider, add the following line at the top of your DAG file:

Reviewer: ```suggestion
Before you can use the GreatExpectationsOperator, you need to import it in your DAG. To import the Great Expectations provider in a given DAG, add the following line to the top of the DAG file in your `dags` directory:
```

> `from great_expectations_provider.operators.great_expectations import GreatExpectationsOperator`

Reviewer: ```suggestion
```py
from great_expectations_provider.operators.great_expectations import GreatExpectationsOperator
```
```

Reviewer: Another butchered suggestion.. GH needs to get smarter!

> Using the operator in a DAG is as simple as defining an instance of the `GreatExpectationsOperator` class and assigning it to a variable, like so:

Reviewer: ```suggestion
To use the operator in the DAG, define an instance of the `GreatExpectationsOperator` class and assign it to a variable. In the following example, we define two different instances of the operator to complete to different steps in a data quality check workflow:
```

```python
ge_data_context_root_dir_with_checkpoint_name_pass = GreatExpectationsOperator(
    task_id="ge_data_context_root_dir_with_checkpoint_name_pass",
    data_context_root_dir=ge_root_dir,
    checkpoint_name="taxi.pass.chk",
)

ge_data_context_config_with_checkpoint_config_pass = GreatExpectationsOperator(
    task_id="ge_data_context_config_with_checkpoint_config_pass",
    data_context_config=example_data_context_config,
    checkpoint_config=example_checkpoint_config,
)
```

> The above code creates two Great Expectations operators assigned to two distinct variables. Once defined, the operators must be put into the DAG, like so:

Reviewer: ```suggestion
Once you define your work through operators, you need to define the order in which your DAG completes the work. To do this, you can define a [relationship](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#relationships). For example, adding the following line to your DAG ensures that your name pass task has to complete before your config pass task can start:
```

```python
ge_data_context_root_dir_with_checkpoint_name_pass >> ge_data_context_config_with_checkpoint_config_pass
```

> The above code defines the DAG as two operations, the first being `ge_data_context_root_dir_with_checkpoint_name_pass` and the second `ge_data_context_config_with_checkpoint_config_pass`, which will run sequentially in that order.

Reviewer: ```suggestion
```

The operator has several optional arguments, but it always requires either a `data_context_root_dir` or a `data_context_config` and either a `checkpoint_name` or `checkpoint_config`.

> The `data_context_root_dir` should point to the `great_expectations` project directory previously generated. When running an Astronomer deployment, it is recommended that this directory is under the `include/` directory, which is automatically generated when creating an Astronomer project. When using a `data_context_config`, the `DataContextConfig` object must also be imported into the project file using:

Reviewer: This info should be up by where we show the operator code, no? Also, not sure what the "`great_expectations` project directory previously generated" is. This is the first I've heard that mentioned. 

It's also getting a bit confusing which setup is Astronomer agnostic and which is not. Like here, we only say where this directory lives for Astronomer, but we don't say where for OSS Airflow.

Author: Yeah, this part is a bit iffy where things really should go. I tried to keep these parameters close to where the operator is defined, maybe it'll be better to split this into a "Parameters" subheading. 

The "`great_expectations` project directory previously generated" is bad wording, but it's a reference to the project files generated when making a new Great Expectations project with the CLI. I can clean that up. But it's a safe assumption as there is a reference at the beginning to having that as a prerequisite to this guide.

Again, thinking I'll just remove the Astronomer related stuff. Maybe at the end we can link to our guide.

> `from great_expectations.data_context.types.base import DataContextConfig`

Reviewer: Let's include all imports together, since that's the natural way someone would write their DAG. We can then break the setup for `DataContextConfig` into its own header, since it's a noticeably different part of the setup.

Author: Well, so the one thing about this is that these imports (for `DataContextConfig` and `CheckpointConfig`) ideally don't go in the DAG, but their own config files. This is currently how they're exemplified in the GE example DAG on their repository (and the DAGs I wrote). I changed this section around a bunch, so might be best to make a decision after I push more changes.

                    "suppress_store_backend_id": True,
                    "base_directory": os.path.join(ge_root_dir, "checkpoints"),
                },
            },
        },
        "expectations_store_name": "expectations_store",
        "validations_store_name": "validations_store",
        "evaluation_parameter_store_name": "evaluation_parameter_store",
        "checkpoint_store_name": "checkpoint_store",
        "data_docs_sites": {
            "local_site": {
                "class_name": "SiteBuilder",
                "show_how_to_buttons": True,
                "store_backend": {
                    "class_name": "TupleFilesystemStoreBackend",
                    "base_directory": os.path.join(
                        ge_root_dir, "uncommitted", "data_docs", "local_site"
                    ),
                },
                "site_index_builder": {"class_name": "DefaultSiteIndexBuilder"},
            }
        },
        "anonymous_usage_statistics": {
            "data_context_id": "abcdabcd-1111-2222-3333-abcdabcdabcd",
            "enabled": False,
        },
        "notebooks": None,
        "concurrency": {"enabled": False},
    }
)
```

> A `checkpoint_name` references a checkpoint in a series of subdirectories relative to the module (which is often the `great_expectations/checkpoints/` path), so that a `checkpoint_name = "taxi.pass.chk"`would reference the file `great_expectations/checkpoints/taxi/pass/chk.yml`. With a `checkpoint_name`, `checkpoint_kwargs` may be passed to the operator to specify additional, overwriting configurations. A `checkpoint_config` may be passed to the operator in place of a name. An example `checkpoint_config` is provided here:

Reviewer: How much of the following code is specific to Airflow? Can we just show the operators and then link out to other GE docs for configuring checkpoint names and contexts?

Author: We can probably link some of this out to other docs, or better yet lines of code in the github repo or registry containing the example DAG this is from. It isn't totally Airflow-specific, but I wanted to include enough example code here that someone could reasonably set up an actual working DAG going through just this one page.

        "evaluation_parameters": {},
        "runtime_configuration": {},
        "validations": [
            {
                "batch_request": {
                    "datasource_name": "my_datasource",
                    "data_connector_name": "default_inferred_data_connector_name",
                    "data_asset_name": "yellow_tripdata_sample_2019-01.csv",
                    "data_connector_query": {"index": -1},
                },
            }
        ],
        "profilers": [],
        "ge_cloud_id": None,
        "expectation_suite_ge_cloud_id": None,
    }
)
```

The `DataContextConfig` and  `CheckpointConfig` objects may be defined in the DAG file, but to avoid excess top-level code, it is recommended these objects be defined in a config file in the `include/` directory and imported into the DAG file.

For a full list of parameters, see the `GreatExpectationsOperator` [documentation](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator). Additional [Great Expectations example DAGs](https://registry.astronomer.io/dags?providers=Great+Expectations&page=1) may be found on the [Astronomer Registry](https://registry.astronomer.io/).

> ## Connections

Reviewer: This seems like it's more about configuring backends than the Airflow connection.

Author: I could change the title to `Connections and Backends`, because really both are needed.

While Airflow alone is a powerful tool, [Astronomer](https://www.astronomer.io/) provides a CLI and cloud platform to bring your DAGs to the next level by providing simple and efficient ways to create Airflow projects, author and publish DAGs, and easily maintain the Kubernetes engine powering DAG runs. To get started quickly with Astronomer’s CLI for use in the steps below, check out the [Astronomer CLI quick start guide](https://docs.astronomer.io/enterprise/cli-quickstart/).

## Install the `GreatExpectationsOperator`

To import the GreatExpectationsOperator in your Airflow project, run the following command to install the Great Expectations provider in your Airflow environment:

```
pip install airflow-provider-great-expectations==0.1.1
```

It’s recommended to specify a version when installing the package. To make use of the latest Great Expectations V3 API, you need to specify a version >= `0.1.0`.

**NOTE:**

> The Great Expectations V3 API requires Airflow 2.0+. If you're still running Airflow 1.x, you need to upgrade to at least 2.0 before using v0.1.0+ of the GreatExpectationsOperator.

Reviewer 2: I think we specify that it needs be greater than 2.1 in the `setup.py` for the Operator.

## Using the `GreatExpectationsOperator`

Before you can use the `GreatExpectationsOperator`, you need to import it in your DAG. You may also need to import the `DataContextConfig`, `CheckpointConfig`, or `BatchRequest` classes as well, depending on how you're using the operator. To import the Great Expectations provider and config and batch classes in a given DAG, add the following line to the top of the DAG file in your `dags` directory:

```python
from great_expectations_provider.operators.great_expectations import GreatExpectationsOperator
from great_expectations.core.batch import BatchRequest
from great_expectations.data_context.types.base import (
    DataContextConfig,
    CheckpointConfig
)
```

> To use the operator in the DAG, define an instance of the `GreatExpectationsOperator` class and assign it to a variable. In the following example, we define two different instances of the operator to complete to different steps in a data quality check workflow:

Reviewer 2: ```suggestion
To use the operator in the DAG, define an instance of the `GreatExpectationsOperator` class and assign it to a variable. In the following example, we define two different instances of the operator to complete two different steps in a data quality check workflow:
```

```python
ge_data_context_root_dir_with_checkpoint_name_pass = GreatExpectationsOperator(
    task_id="ge_data_context_root_dir_with_checkpoint_name_pass",
    data_context_root_dir=ge_root_dir,
    checkpoint_name="taxi.pass.chk",
)

ge_data_context_config_with_checkpoint_config_pass = GreatExpectationsOperator(
    task_id="ge_data_context_config_with_checkpoint_config_pass",
    data_context_config=example_data_context_config,
    checkpoint_config=example_checkpoint_config,
)
```

Once you define your work through operators, you need to define the order in which your DAG completes the work. To do this, you can define a [relationship](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#relationships). For example, adding the following line to your DAG ensures that your name pass task has to complete before your config pass task can start:

```python
ge_data_context_root_dir_with_checkpoint_name_pass >> ge_data_context_config_with_checkpoint_config_pass
```

### Operator Parameters

> The operator has several optional parameters, but it always requires either a `data_context_root_dir` or a `data_context_config` and either a `checkpoint_name` or `checkpoint_config`.

Reviewer 2: Even though users are supposed to have a DataContext and Checkpoint configured as prereqs for this guide, I wonder if it makes sense to have some simple definitions here as explanation. Something like:
A [Data Context](https://docs.greatexpectations.io/docs/reference/data_context) represents a Great Expectations project. It organizes storage and access for Expectation Suites, Datasources, notification settings, and data fixtures.

[Checkpoints](https://docs.greatexpectations.io/docs/reference/checkpoints_and_actions) provide a convenient abstraction for bundling the validation of a Batch (or Batches) of data against an Expectation Suite (or several), as well as the actions that should be taken after the validation. 

I'm think of a situation where a user comes to this doc from Airflow without much knowledge of GE and is trying to get a quick glimpse of the basic building blocks of running the Operator. It might even make sense to have these explanations near the start of the doc.

> The `data_context_root_dir` should point to the `great_expectations` project directory generated when you created the project with the CLI. If using a `data_context_config`, a `DataContextConfig` must be defined, as in [this example](https://github.com/great-expectations/airflow-provider-great-expectations/blob/main/include/great_expectations/object_configs/example_data_context_config.py).

Reviewer 2: ```suggestion
The `data_context_root_dir` should point to the `great_expectations` project directory generated when you created the project with the CLI. If using an in-memory `data_context_config`, a `DataContextConfig` must be defined, as in [this example](https://github.com/great-expectations/airflow-provider-great-expectations/blob/main/include/great_expectations/object_configs/example_data_context_config.py).
```

> A `checkpoint_name` references a checkpoint in a series of subdirectories relative to the module (which is often the `great_expectations/checkpoints/` path), so that a `checkpoint_name = "taxi.pass.chk"` would reference the file `great_expectations/checkpoints/taxi/pass/chk.yml`. With a `checkpoint_name`, `checkpoint_kwargs` may be passed to the operator to specify additional, overwriting configurations. A `checkpoint_config` may be passed to the operator in place of a name, and can be defined like [this example](https://github.com/great-expectations/airflow-provider-great-expectations/blob/main/include/great_expectations/object_configs/example_checkpoint_config.py).

Reviewer 2: ```suggestion
A `checkpoint_name` references a checkpoint in the project CheckpointStore defined in the DataContext (which is often the `great_expectations/checkpoints/` path), so that a `checkpoint_name = "taxi.pass.chk"` would reference the file `great_expectations/checkpoints/taxi/pass/chk.yml`. With a `checkpoint_name`, `checkpoint_kwargs` may be passed to the operator to specify additional, overwriting configurations. A `checkpoint_config` may be passed to the operator in place of a name, and can be defined like [this example](https://github.com/great-expectations/airflow-provider-great-expectations/blob/main/include/great_expectations/object_configs/example_checkpoint_config.py).
```

> For a full list of parameters, see the `GreatExpectationsOperator` [documentation](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator).

Reviewer 2: Just checking, it looks like this is not yet updated. I just want to make sure this is updated before we merge the doc.

Author: Yep, I'll make sure it's updated today.

Author: It should be updated, seems to be taking the right parameters and pointing to the right github repo.

### Connections and Backends

> The `GreatExpectationsOperator` can run a checkpoint on a dataset stored in any backend compatible with Great Expectations. All that’s needed to get the Operator to point at an external dataset is to set up an [Airflow Connection](https://www.astronomer.io/guides/connections) to the datasource, and add the connection to your Great Expectations project, e.g. [using the CLI to add an Athena backend](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/database/athena). Then, if using a `DataContextConfig` or `CheckpointConfig`, ensure that the `"datasources"` field refers to your backend connection name.

Reviewer 2: A couple of notes on this:
- I don't think an Airflow Connection is needed here (and I'm not sure how it could be used). I believe that it's just the Datasource that has to be specified.
- Instead of Athena (which is somewhat experimental), it would probably be better to link to our guide to Postgres instead (https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/database/postgres). BigQuery, Snowflake, or Redshift would also be fine.
- I'm trying to think of how to make this as simple as possible. Something like
```suggestion
The `GreatExpectationsOperator` can run a checkpoint on data stored in any backend compatible with Great Expectations. All that’s needed to get the Operator to point at an external dataset is to configure and save a Great Expectations Datasource, which you can do with the CLI, (e.g. [with PostgreSQL](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/database/postgres)), or [in-memory](https://docs.greatexpectations.io/docs/guides/setup/configuring_data_contexts/how_to_instantiate_a_data_context_without_a_yml_file). Then, once you [create your ExpectationSuite](https://docs.greatexpectations.io/docs/guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data) and [Checkpoint](https://docs.greatexpectations.io/docs/guides/validation/checkpoints/how_to_create_a_new_checkpoint), you'll be in business!
```

Author: You know, I haven't actually tried connecting to an external datasource without a connection configured... I just assumed Airflow would need it to make the connection. If that's the case, then might as well just remove the connections section all together, because that's GE-specific.
